### PR TITLE
DPDK pps testcases

### PR DESCRIPTION
**Pull request by justin-parus**

### CODE REVIEW

File: Testscripts/Windows/dpdk-testPMDmulticore.ps1 [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+    [int]$number = $null
+
+    if ([int32]::TryParse($candidate, [ref]$number)) {
+        return $number
+    } else {
+        throw "Failed to cast str to int"
+    }
+}
+
+function verifyPerf() {
```
Comment (juhlee-microsoft) [inline Testscripts/Windows/dpdk-testPMDmulticore.ps1]: Add try-catch-finally structure if you can not get return value from source.

File: Testscripts/Windows/dpdk-testPMDmulticore.ps1 [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+    # use temp so when a case fails we still check the rest
+    $tempResult = "PASS"
+    $testpmdDataCsv = Import-Csv -Path $LogDir\dpdkTestPmd.csv
+
+    $allPpsData = [Xml](Get-Content .\XML\Other\pps_lowerbound.xml)
+    $sizeData = Select-Xml -Xml $allPpsData -XPath "testpmdpps/$vmSize" | Select-Object -ExpandProperty Node
+
+    if ($null -eq $sizeData) {
+        $msg = "No pps data for VM size $vmSize"
+        LogMsg $msg
```
Comment (juhlee-microsoft) [inline Testscripts/Windows/dpdk-testPMDmulticore.ps1]: From here to down there, if it is about error or failure, you will change to LogError or LogWarn.  No one can pay attention to 'LogMsg with Perf Failure'

File: Testscripts/Windows/dpdk-testPMDmulticore.ps1 [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+            $msg = "No pps data for test mode $($testRun.TestMode)"
+            LogMsg $msg
+            throw $msg
+        }
+    }
+
+    return $tempResult
+}
+
+function prepareParameters() {
```
Comment (juhlee-microsoft) [inline Testscripts/Windows/dpdk-testPMDmulticore.ps1]: Same comment applied - try-catch-finally, LogMsg to LogWarn/LogError.

File: Testscripts/Windows/dpdk-testPMDmulticore.ps1 [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+    if ($vmSizes.count -ne 2) {
+        $msg = "Test only supports two VMs"
+        LogMsg $msg
+        throw $msg
+    }
+
+    if ($vmSizes[0] -ne $vmSizes[1]) {
+        $msg = "Test only supports VMs of same size"
+        LogMsg $msg
+        throw $msg
```
Comment (juhlee-microsoft) [inline Testscripts/Windows/dpdk-testPMDmulticore.ps1]: If this throw happens, where would run-time go?

File: Testscripts/Windows/DPDK-TEMPLATE.ps1 [hunk 5]
```diff
@@ -0,0 +1,140 @@
+# Copyright (c) Microsoft Corporation. All rights reserved.
```
Comment (juhlee-microsoft) [inline Testscripts/Windows/DPDK-TEMPLATE.ps1]: Can you rename the filename? It looks report generation. If so, the name 'template' makes confusion.

File: Testscripts/Linux/dpdkUtils.sh [hunk 6]
```diff
@@ -0,0 +1,281 @@
...
+        distro_version=`cat /etc/oracle-release | sed s/.*release\ // | sed s/\ .*//`
+    elif [ -f /etc/redhat-release ] ; then
+        distro_version=`cat /etc/redhat-release | sed s/.*release\ // | sed s/\ .*//`
+    elif [ -f /etc/os-release ] ; then
+        distro_version=`cat /etc/os-release|sed 's/"//g'|grep "VERSION_ID="| sed 's/VERSION_ID=//'| sed 's/\r//'`
+    fi
+    echo $distro_version
+}
+
+function detectDistro() {
```
Comment (juhlee-microsoft) [inline Testscripts/Linux/dpdkUtils.sh]: duplicated

File: Testscripts/Windows/dpdk-testPMDmulticore.ps1 (hunk 4 above)
Comment (justin-parus) [reply Testscripts/Windows/dpdk-testPMDmulticore.ps1]: DPDK-TEMPLATE which I calls these functions has a try-catch-finally block where the error will be reported and test will be aborted

File: XML/TestCases/PerformanceTests-DPDK.xml [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+        <TestParameters>
+            <param>dpdkSrcLink="https://fast.dpdk.org/rel/dpdk-18.05.tar.xz"</param>
+            <param>testDuration=120</param>
+            <param>modes="rxonly io"</param>
+            <param>cores="2 4 8 16"</param>
+        </TestParameters>
+        <Platform>Azure</Platform>
+        <Category>Performance</Category>
+        <Area>DPDK</Area>
+        <Tags>dpdk,pps,testpmd,cores,multicore,network,hv_netvsc,sriov</Tags>
```
[the remark this answers is not in the document]
Comment (justin-parus) [reply XML/TestCases/PerformanceTests-DPDK.xml]: Tags removed

File: Testscripts/Windows/dpdk-testPMDmulticore.ps1 (hunk 3 above)
Comment (justin-parus) [reply Testscripts/Windows/dpdk-testPMDmulticore.ps1]: These errors are designed to leave this function, and be reported by DPDK-TEMPLATE

File: Testscripts/Windows/DPDK-TEMPLATE.ps1 (hunk 5 above)
Comment (justin-parus) [reply Testscripts/Windows/DPDK-TEMPLATE.ps1]: does `DPDK-TESTCASE-DRIVER.ps1` work?

Comment (juhlee-microsoft) [reply Testscripts/Windows/DPDK-TEMPLATE.ps1]: ok.

File: Testscripts/Windows/dpdk-testPMDmulticore.ps1 (hunk 3 above)
Comment (juhlee-microsoft) [reply Testscripts/Windows/dpdk-testPMDmulticore.ps1]: As long as the exception is handled in DPDK-Test-case-driver, it is ok. Close this comment.

File: Testscripts/Windows/dpdk-testPMDmulticore.ps1 (hunk 4 above)
Comment (juhlee-microsoft) [reply Testscripts/Windows/dpdk-testPMDmulticore.ps1]: acked.

File: XML/TestCases/PerformanceTests-DPDK.xml [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+        <Tags>dpdk,pps,testpmd,network,hv_netvsc,sriov</Tags>
+    </test>	
+    <test>
+        <testName>PERF-DPDK-MULTICORE-PPS-DS15</testName>
+        <testScript>DPDK-TEMPLATE.ps1</testScript>
+        <setupType>DS15v2_2NIC_x2</setupType>
+        <AdditionalHWConfig>
+            <Networking>SRIOV</Networking>
+        </AdditionalHWConfig>        
+        <files>.\Testscripts\Linux\dpdk-testPMDmulticore.sh,.\Testscripts\Windows\dpdk-testPMDmulticore.ps1</files>
```
Comment (madhan-s) [inline XML/TestCases/PerformanceTests-DPDK.xml]: Rename files as discussed

File: Testscripts/Windows/dpdk-testPMDmulticore.ps1 [hunk 9]
```diff
@@ -0,0 +1,144 @@
...
+            $tempResult = "FAIL"
+        }
+
+        if ($testRun.TestMode -eq "rxonly") {
+            LogMsg "compare rx pps $($testRun.RxPps) with lowerbound $($coreData.rx)"
+            if ((castStrToInt($testRun.RxPps)) -lt (castStrToInt($coreData.rx))) {
+                LogMsg "Perf Failure in $($testRun.TestMode) mode; $($testRun.RxPps) must be > $($coreData.rx)"
+                $tempResult = "FAIL"
+            }
+        } elseif ($testRun.TestMode -eq "io") {
```
Comment (madhan-s) [inline Testscripts/Windows/dpdk-testPMDmulticore.ps1]: please change the mode to fwd

File: Testscripts/Windows/dpdk-testPMDmulticore.ps1 [hunk 10]
```diff
@@ -0,0 +1,144 @@
...
+            LogMsg "compare rx pps $($testRun.RxPps) with lowerbound $($coreData.fwdrx)"
+            LogMsg "compare fwdtx pps $($testRun.ReTxPps) with lowerbound $($coreData.fwdtx)"
```
Comment (madhan-s) [inline Testscripts/Windows/dpdk-testPMDmulticore.ps1]: Change to fwdtx

File: Testscripts/Linux/dpdkUtils.sh [hunk 11]
```diff
@@ -0,0 +1,281 @@
...
+
+# gets synthetic vf pairs by comparing MAC addresses.
+#   will ignore the default route interface even if it has accelerated networking,
+#   which should be the primaryNIC with pubilc ip to which you SSH
+# recommend to capture output in array like so
+#   pairs=($(getSyntheticVfPair))
+#   then synthetic ${pairs[n]} maps to vf pci address ${pairs[n+1]}
+#   when starting from zero i.e. index 1 and 2 have no relation
+#   if captured output is empty then no VFs exist
+function getSyntheticVfPairs() {
```
Comment (madhan-s) [inline Testscripts/Linux/dpdkUtils.sh]: Move this to a common file.

File: Testscripts/Linux/dpdkUtils.sh [hunk 12]
```diff
@@ -0,0 +1,281 @@
...
+        fi
+    elif [ "$linux_distribution" == "ol" ]; then
+        linux_distribution='oracle'
+    elif echo "$linux_distribution" | grep -qi "debian"; then
+        linux_distribution='debian'
+    fi
+    echo "$(echo "$linux_distribution" | awk '{print tolower($0)}')"
+}
+
+function sourceScript() {
```
Comment (madhan-s) [inline Testscripts/Linux/dpdkUtils.sh]: Move this to a common file.

File: XML/TestCases/PerformanceTests-DPDK.xml (hunk 8 above)
Comment (justin-parus) [reply XML/TestCases/PerformanceTests-DPDK.xml]: Done

File: Testscripts/Linux/dpdkUtils.sh [hunk 13]
```diff
@@ -0,0 +1,238 @@
...
+        if [ $res -eq 1 ]; then
+            LogErr "ERROR: must pass valid ip to modprobeSetup"
+            SetTestStateAborted
+            exit 1
+        fi
+        ssh ${1} "${modprobeCMD}"
+    fi
+}
+
+function sourceScript() {
```
Comment (lizzha) [inline Testscripts/Linux/dpdkUtils.sh]: this function can also be moved to the common utils.sh

File: Testscripts/Linux/dpdk-testPMD.sh [hunk 14]
```diff
@@ -0,0 +1,176 @@
...
+#
+# dpdk-testPMDmulticore.sh
+# Description:
+#	This script runs testpmd in various modes scaling across various cores.
+# 	It places testpmd output in $LOGDIR, and the parses output to get avg pps
+# 	numbers. The accompanying ps1 script makes sure testpmd performs above the
+#	required threshold.
+#
+#############################################################################
+
```
Comment (lizzha) [inline Testscripts/Linux/dpdk-testPMD.sh]: Could you use consistent naming conventions in all the bash scripts?
I see several different type of naming style used.
We can refer to https://google.github.io/styleguide/shell.xml#Naming_Conventions, basically it's lower_case for function can variable names

File: Testscripts/Linux/dpdk-testPMD.sh [hunk 15]
```diff
@@ -0,0 +1,176 @@
...
+#
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the Apache License.
+
+#############################################################################
+#
+# dpdk-testPMDmulticore.sh
+# Description:
+#	This script runs testpmd in various modes scaling across various cores.
+# 	It places testpmd output in $LOGDIR, and the parses output to get avg pps
```
Comment (ader1990) [inline Testscripts/Linux/dpdk-testPMD.sh]: the -> then

File: Testscripts/Linux/dpdk-testPMD.sh [hunk 16]
```diff
@@ -0,0 +1,176 @@
...
+        exit 1
+    fi
+
+    iface="${pairs[0]}"
+    busaddr="${pairs[1]}"
+
+    for testmode in $modes; do
+        LogMsg "Ensuring free hugepages"
+        freeHugeCMD="rm -rf /dev/hugepages/*"
+        ssh ${server} $freeHugeCMD
```
Comment (ader1990) [inline Testscripts/Linux/dpdk-testPMD.sh]: isn t it required to disable host checking, as it might hang forever

Comment (justin-parus) [reply Testscripts/Linux/dpdk-testPMD.sh]: Correct, `ProvisionVMsForLisa` handles this. Added note

**Comment by justin-parus** [comment]
The bash style guide linked above recommends spaces while the LISA guide recommends tabs. I have elected to use the LISA guide for the powershell scripts and the bash guide for the bash scripts. Please let me know if I should change this!

**Comment by justin-parus** [comment]
Please review when you get a chance

File: Testscripts/Linux/dpdk-testPMD.sh (hunk 14 above)
Comment (justin-parus) [reply Testscripts/Linux/dpdk-testPMD.sh]: Fixed

File: Testscripts/Windows/dpdk-testPMD.ps1 [hunk 17]
```diff
@@ -0,0 +1,119 @@
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the Apache License.
+
+function verifyPerf() {
```
Comment (lizzha) [inline Testscripts/Windows/dpdk-testPMD.ps1]: use Verb-Entity naming

**Pull request merged**